### PR TITLE
[5.4] add OverlappingStrategy for Schedule/Event

### PR DESCRIPTION
**Pull request by fetzi**
The `withoutOverlapping` implementation in the `Event` class is not working in all cases. For example when the process gets killed, the cache entry is not removed and the scheduled command will not start again until the cache entry expires.

Therefore I've implemented an `OverlappingStrategy` Interface to be able to override the standard cache strategy. The interface specifies 3 methods:
* **prevent**: prevent overlapping for the current event's command (e.g. writes cache entry)
* **reset**: resets the overlapping strategy for the event's command (e.g. deletes cache entry)
* **overlaps**: checks if the event's command is overlapping with an already running command (e.g. checks if cache entry exists)

To be compatible with the current implementation I've added a `CacheOverlappingStrategy` that provides the same functionality as the current implementation.

### CODE REVIEW

**Comment by taylorotwell** [comment]
So this doesn't actually change the current behavior? What does your custom strategy look like?

**Comment by fetzi** [comment]
I check if the artisan command is running already by filtering the output if `ps aux`, I know that this solution is only valid for unix operating systems and therefore the current behavior should be kept.

But I can include my custom strategy in the PR if you want me to?

File: src/Illuminate/Console/Scheduling/CacheOverlappingStrategy.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    private $cache;
+
+    public function __construct(Cache $cache)
+    {
+        $this->cache = $cache;
+    }
+
+    public function prevent(Event $event)
+    {
+        $this->cache->put($event->mutexName(), true, 1440);
```
Comment (alain-lf) [inline src/Illuminate/Console/Scheduling/CacheOverlappingStrategy.php]: Would be nice to be able to modify this number in configs.

1440 = 24 hours.

Depending on your workload, if one of your servers goes down, you might not want to wait a full day for the key to expire and resume schedule.

**Review by alain-lf — CHANGES_REQUESTED** [review]
Would be nice to be able to modify the expiration of the mutex in configs

**Comment by fetzi** [comment]
Adding the cache entry lifetime for the overlapping check to the config would be an option, but I think this would be to much change for a minor feature change. 

And also a shorter cache entry lifetime does not solve the overlapping problem when the process dies, it only reduces the time until the process can be restarted. Maybe the `ProcessOverlappingStrategy` would be a solution fror you as well?

@taylorotwell what do you think about the config for the mutex cache entry liftetime?

**Comment by taylorotwell** [comment]
You wouldn't really want a global config item for this. It would be job specific.

**Comment by fetzi** [comment]
That's true, I will skip this for the PR because it is targeting the extensibility of the overlapping strategy. Is the PR ready for merging or do you need additional infos?

**Comment by fetzi** [comment]
@taylorotwell ?

**Comment by taylorotwell** [comment]
No need to keep pinging. I will eventually look at it.

**Pull request merged**